Frameworks/Foundation/NSFileHandle.mm

```
@@ -175,37 +121,103 @@ + (instancetype)fileHandleForUpdatingAtPath:(NSString*)file {
+    return nullDeviceHandle;
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithFileDescriptor:(int)fileDescriptor closeOnDealloc:(BOOL)flag {
+    if (self = [super init]) {
+        _fileDescriptor = fileDescriptor;
+        _closeOnDealloc = flag;
+    }
+
+    return self;
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithFileDescriptor:(int)fileDescriptor {
+    return [self initWithFileDescriptor:fileDescriptor closeOnDealloc:NO];
+}
+
+- (instancetype)_initWithFileAtPath:(NSString*)file openType:(_NSFileOpenMode)type {
+    if (file == nil) {
+        [self release];
+        return nil;
+    }
+    if (self = [self initWithFileDescriptor:-1 closeOnDealloc:YES]) {
+        _fileDescriptor = [self _openFile:file openType:type];
+
+        if (_fileDescriptor == -1) {
+            TraceError(L"NSFileHandle", L"Unable to open file, errno:%d", errno);
+            [self release];
+            return nil;
+        }
+
+        if (type == _NSFileOpenModeUpdate) {
+            [self seekToFileOffset:0];
+        }
+    }
+
+    return self;
+}
+
+/**
+ @Status Interoperable
+*/
+- (void)dealloc {
+    if (_closeOnDealloc) {
+        [self closeFile];
+    }
+    [super dealloc];
+}
+
+- (int)_openFile:(NSString*)file openType:(_NSFileOpenMode)type {
+    switch (type) {
+        case _NSFileOpenModeRead:
+            return EbrOpenWithPermission((char*)([file UTF8String]), O_RDONLY | _O_BINARY, _SH_DENYWR, _S_IREAD);
+        case _NSFileOpenModeWrite:
+            return EbrOpenWithPermission((char*)([file UTF8String]), _O_WRONLY | _O_BINARY | _O_CREAT, _SH_DENYNO, _S_IREAD | _S_IWRITE);
+        case _NSFileOpenModeUpdate:
+            return EbrOpenWithPermission((char*)([file UTF8String]),
+                                         O_RDWR | _O_CREAT | O_APPEND | _O_BINARY,
+                                         _SH_DENYNO,
+                                         _S_IREAD | _S_IWRITE);
+        default:
+            break;
```

Since this is internal, should we fail if we get an invalid type?